### PR TITLE
Fix Tiling in `ConsistencyDecoderVAE`

### LINKED ISSUE #6126: AttributeError: 'ConsistencyDecoderVAE' object has no attribute 'tile_sample_min_size'
### Describe the bug

The new decoder has enable_tiling and disable_tiling methods but has not correctly implemented the support attributes.

### Reproduction

```python
import torch
from diffusers import ConsistencyDecoderVAE
vae = ConsistencyDecoderVAE.from_pretrained("openai/consistency-decoder")
vae.enable_tiling()
vae.encode(torch.randn(1, 4, 256, 256))
```

### Logs

```shell
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <cell line: 1>                                                                                │
│                                                                                                  │
│ D:\CondaM\lib\site-packages\diffusers\utils\accelerate_utils.py:46 in wrapper                    │
│                                                                                                  │
│   43 │   def wrapper(self, *args, **kwargs):                                                     │
│   44 │   │   if hasattr(self, "_hf_hook") and hasattr(self._hf_hook, "pre_forward"):             │
│   45 │   │   │   self._hf_hook.pre_forward(self)                                                 │
│ ❱ 46 │   │   return method(self, *args, **kwargs)                                                │
│   47 │                                                                                           │
│   48 │   return wrapper                                                                          │
│   49                                                                                             │
│                                                                                                  │
│ D:\CondaM\lib\site-packages\diffusers\models\consistency_decoder_vae.py:285 in encode            │
│                                                                                                  │
│   282 │   │   │   │   [`~models.consistency_decoder_vae.ConsistencyDecoderVAEOutput`] is retur   │
│   283 │   │   │   │   is returned.                                                               │
│   284 │   │   """                                                                                │
│ ❱ 285 │   │   if self.use_tiling and (x.shape[-1] > self.tile_sample_min_size or x.shape[-2] >   │
│   286 │   │   │   return self.tiled_encode(x, return_dict=return_dict)                           │
│   287 │   │                                                                                      │
│   288 │   │   if self.use_slicing and x.shape[0] > 1:                                            │
│                                                                                                  │
│ D:\CondaM\lib\site-packages\diffusers\models\modeling_utils.py:222 in __getattr__                │
│                                                                                                  │
│    219 │   │   │   return self._internal_dict[name]                                              │
│    220 │   │                                                                                     │
│    221 │   │   # call PyTorch's https://pytorch.org/docs/stable/_modules/torch/nn/modules/modul  │
│ ❱  222 │   │   return super().__getattr__(name)                                                  │
│    223 │                                                                                         │
│    224 │   @property                                                                             │
│    225 │   def is_gradient_checkpointing(self) -> bool:                                          │
│                                                                                                  │
│ D:\CondaM\lib\site-packages\torch\nn\modules\module.py:1614 in __getattr__                       │
│                                                                                                  │
│   1611 │   │   │   modules = self.__dict__['_modules']                                           │
│   1612 │   │   │   if name in modules:                                                           │
│   1613 │   │   │   │   return modules[name]                                                      │
│ ❱ 1614 │   │   raise AttributeError("'{}' object has no attribute '{}'".format(                  │
│   1615 │   │   │   type(self).__name__, name))                                                   │
│   1616 │                                                                                         │
│   1617 │   def __setattr__(self, name: str, value: Union[Tensor, 'Module']) -> None:             │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'ConsistencyDecoderVAE' object has no attribute 'tile_sample_min_size'
```
```


### System Info

- `diffusers` version: 0.24.0
- Platform: Windows-10-10.0.22621-SP0
- Python version: 3.8.13
- PyTorch version (GPU?): 2.0.1+cu118 (True)
- Huggingface_hub version: 0.19.4
- Transformers version: 4.29.0
- Accelerate version: 0.18.0
- xFormers version: not installed
- Using GPU in script?: Y/N both the same issue
- Using distributed or parallel set-up in script?: N

### Who can help?

@patrickvonplaten @saya

### PR DESCRIPTION
**Pull request by standardAI**
Fixes #6126.

This pull request includes three commits:

- Fix tiling

- Add docstring to `decode` method in `ConsistencyDecoderVAE`

- Fix typos

I tested with this:
```py
import torch
from diffusers import ConsistencyDecoderVAE

vae = ConsistencyDecoderVAE.from_pretrained("openai/consistency-decoder").to("cuda")
vae.enable_tiling()
vae.encode(torch.randn((1, 3, 256, 256), device="cuda"))
```
Also, this PR passes `./tests/models/autoencoders/test_models_vae.py` tests in my local setup.

@yiyixuxu @sayakpaul

### CODE REVIEW

**Review by sayakpaul — COMMENTED** [review]
This is looking solid. 

I left a major comment on how we should expose the tiling arguments (init vs. via `enable_vae_tiling()`). 

Can we also add a test for this?

File: src/diffusers/models/autoencoders/consistency_decoder_vae.py [hunk 1]
```diff
@@ -153,6 +155,16 @@ def __init__(
...
 
+        # only relevant if vae tiling is enabled
+        self.tile_sample_min_size = self.config.sample_size
+        sample_size = (
+            self.config.sample_size[0]
+            if isinstance(self.config.sample_size, (list, tuple))
+            else self.config.sample_size
+        )
+        self.tile_latent_min_size = int(sample_size / (2 ** (len(self.config.block_out_channels) - 1)))
+        self.tile_overlap_factor = 0.25
```
[the remark this answers is not in the document]
Comment (standardAI) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: I added configurability to `enable_vae_tiling()`. If this is confirmed I will equalize the copied places, and add a test(s).

[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: @yiyixuxu WDYT?

File: src/diffusers/models/autoencoders/consistency_decoder_vae.py [hunk 2]
```diff
@@ -272,7 +284,7 @@ def encode(
         Args:
             x (`torch.FloatTensor`): Input batch of images.
             return_dict (`bool`, *optional*, defaults to `True`):
-                Whether to return a [`~models.consistecy_decoder_vae.ConsistencyDecoderOoutput`] instead of a plain
+                Whether to return a [`~models.consistency_decoder_vae.ConsistencyDecoderVAEOutput`] instead of a plain
```
[the remark this answers is not in the document]
Comment (standardAI) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:293]: I am just obsessed with typos, sorry 😅.

File: src/diffusers/models/autoencoders/consistency_decoder_vae.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: I haven't looked into the vae tiling algorithem enough to have an opinion if it is meaningful for users to experiment with these config or not. I don't think we have seen any requests for that 

in any case, I think we can just keep it consistent with `AutoencoderKL` in this PR  - feel free to open a separate PR to refactor both if you think it makes sense (and would be nice to explain a little bit the use cases when and how they want to configure them too)

**Review by yiyixuxu — COMMENTED** [review]
hey! thanks for the PR.
it's looking great!

I left a feedback and additionally, could you add a test_vae_tiling test to ./tests/models/autoencoders/test_models_vae.py( similar to https://github.com/huggingface/diffusers/blob/d3986f18beac69b2ecb0af80b0e6f3134cc5419d/tests/pipelines/test_pipelines_common.py#L101)

File: src/diffusers/models/autoencoders/consistency_decoder_vae.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: Sure. From [our docs](https://huggingface.co/docs/diffusers/main/en/optimization/memory#tiled-vae):

> Tiled VAE processing also enables working with large images on limited VRAM (for example, generating 4k images on 8GB of VRAM) by splitting the image into overlapping tiles, decoding the tiles, and then blending the outputs together to compose the final image. 

You might have seen some discussions and issues about "smooth blending" of multiple tiles for a single high-resolution image.

[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: @sayakpaul 
feel free to open a PR later to refactor all the vae decoders! and make a nice doc (or at least doc string)

in particular, we need to explain on high level when and how `tile_latent_min_size` and `tile_overlap_factor` should be configured (currently they are not configurable and calculated based on the `sample_size`)

[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: Will do upon community interest. So far, there hasn't been much of an interest around it.

[the remark this answers is not in the document]
Comment (standardAI) [reply src/diffusers/models/autoencoders/consistency_decoder_vae.py:166]: Reverted configurability.

**Review by sayakpaul — COMMENTED** [review]
Looking good! Let's add a test as well?

**Comment by yiyixuxu** [comment]
@standardAI let us know when you are ready for a final review

**Comment by standardAI** [comment]
ASAP.

**Comment by standardAI** [comment]
It seems to pass the test in Colab.

**Review by yiyixuxu — APPROVED** [review]

File: tests/models/autoencoders/test_models_vae.py [hunk 3]
```diff
@@ -1116,3 +1116,33 @@ def test_sd_f16(self):
...
+            generator=torch.Generator("cpu").manual_seed(0),
+        ).images[0]
+
+        assert torch_all_close(out_1, out_2, atol=5e-3)
+
+        # test that tiled decode works with various shapes
+        shapes = [(1, 4, 73, 97), (1, 4, 97, 73), (1, 4, 49, 65), (1, 4, 65, 49)]
+        for shape in shapes:
+            image = torch.zeros(shape, device=torch_device)
+            pipe.vae.decode(image)
```
Comment (sayakpaul) [inline tests/models/autoencoders/test_models_vae.py:1148]: Would be nice to isolate this as a separate test. That's easier for us to look into later.

**Review by sayakpaul — COMMENTED** [review]
One comment and once addressed, we can ship!

File: tests/models/autoencoders/test_models_vae.py (hunk 3 above)
Comment (standardAI) [reply tests/models/autoencoders/test_models_vae.py:1148]: The testing example in [YiYi's reply](https://github.com/huggingface/diffusers/pull/7290#pullrequestreview-1935513923) controls shapes too:
https://github.com/huggingface/diffusers/blob/d3986f18beac69b2ecb0af80b0e6f3134cc5419d/tests/pipelines/test_pipelines_common.py#L101
Should it be separated there as well?

Comment (sayakpaul) [reply tests/models/autoencoders/test_models_vae.py:1148]: Probably a future PR then.

**Comment by sayakpaul** [comment]
Please ping me once the CI run is done.

**Comment by standardAI** [comment]
> Please ping me once the CI run is done.

@sayakpaul

**Comment by sayakpaul** [comment]
Thank you for this impressive contribution!

**Comment by standardAI** [comment]
Thanks so much for reviewing and merging!

**Pull request merged**